utils/net_utils/src/download.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use async_std::io::BufRead;
+use futures::prelude::*;
+use isahc::{Body, HttpClient};
+use pbr::ProgressBar;
+use pin_project_lite::pin_project;
+use std::io::{self, Read, Result as IOResult, Stdout, Write};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use thiserror::Error;
+use url::Url;
+
+#[derive(Debug, Error)]
+enum DownloadError {
+    #[error("Cannot read a file header")]
+    HeaderError,
+}
+
+pin_project! {
+    pub struct FetchProgress<R, W: Write> {
+        #[pin]
+        pub inner: R,
+        pub progress_bar: ProgressBar<W>,
+    }
+}
+
+impl<R, W: Write> FetchProgress<R, W> {
+    pub fn finish(&mut self) {
+        self.progress_bar.finish();
+    }
+}
+
+impl<R: AsyncRead + Unpin, W: Write> AsyncRead for FetchProgress<R, W> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let r = Pin::new(&mut self.inner).poll_read(cx, buf);
+        if let Poll::Ready(Ok(size)) = r {
+            self.progress_bar.add(size as u64);
+        }
+        r
+    }
+}
+
+impl<R: BufRead + Unpin, W: Write> BufRead for FetchProgress<R, W> {
+    fn poll_fill_buf(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<IOResult<&'_ [u8]>> {
+        let this = self.project();
+        this.inner.poll_fill_buf(cx)
+    }
+
+    fn consume(mut self: Pin<&mut Self>, amt: usize) {
+        Pin::new(&mut self.inner).consume(amt)
+    }
+}
+
+impl<R: Read, W: Write> Read for FetchProgress<R, W> {
+    fn read(&mut self, buf: &mut [u8]) -> IOResult<usize> {
+        self.inner.read(buf).map(|n| {
+            self.progress_bar.add(n as u64);
+            n
+        })
+    }
+}
+
+/// Builds Reader for a provided URL.
+pub fn make_reader(
```

Comment:
can you rename this function? This won't work in a generalized case and can only be used for car files over the network, perhaps `fetch_car_file` or something?

Reply:
my intent was to make it usable in a generalized case - anything else is missed you think?

Reply:
ah, my apologies, I misread something. Optionally can just indicate that the reader is coming from an http request, since `make_reader` can be ambiguous to any type of reader (possibly confusing with a file reader since it's only a string param)

Reply:
yeah. Renamed it to `make_http_reader` and passing `url: Url` instead of `&str` to minimize the confusion